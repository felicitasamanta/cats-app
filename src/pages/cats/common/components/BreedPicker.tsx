import React from "react";
import { useQueryParams } from "@/common/hooks/useQueryParams";
import { Dropdown, Options, Option } from "@/common/components/Dropdown";
import { useBreeds } from "../hooks/useBreeds";

type Props = {
  isLoading: boolean;
};

const BreedPicker: React.FC<Props> = ({ isLoading }) => {
  const { isLoading: isBreedsLoading, breeds } = useBreeds();
  const { params, setQueryParams } = useQueryParams();
  const { breed_ids: breedId } = params;

  const options: Options =
    breeds?.map(({ id, name }) => ({
      name,
      value: id,
      selected: id === breedId,
    })) || [];

  options.unshift({
    name: "All",
    value: "all",
  });

  type NewType = Option;

  const onChange = (option: NewType) => {
    setQueryParams({
      ...params,
      breed_ids: option.value === "all" ? null : option.value,
      page: null,
    });
  };

  return (
    <Dropdown
      options={options}
      isLoading={isLoading || isBreedsLoading}
      placeHolder="Breeds"
      preselect
      onChange={onChange}
    />
  );
};

export { BreedPicker };
