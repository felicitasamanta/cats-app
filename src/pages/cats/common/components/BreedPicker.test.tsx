import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from "vitest";
import { BreedPicker } from "./BreedPicker";
import { TestsWrapper } from "@/common/tests/TestsWrapper";
import {
  cleanup,
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { setupServer } from "msw/node";
import { rest } from "msw";

const server = setupServer(
  rest.get("https://api.thecatapi.com/v1/breeds", (_, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json([
        {
          id: "abys",
          name: "Abyssinian",
        },
        {
          id: "aege",
          name: "Aegean",
        },
      ])
    );
  }),

  rest.get("https://api.thecatapi.com/v1/images/search", (_, res, ctx) => {
    return res(ctx.status(200), ctx.json([]));
  })
);

describe("BreedPicker", () => {
  beforeAll(() => {
    server.listen({ onUnhandledRequest: "error" });
  });

  beforeEach(() => {
    vi.resetModules();
  });

  afterEach(() => {
    cleanup();
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  it("should match snapshot", async () => {
    const component = render(<BreedPicker isLoading={false} />, {
      wrapper: TestsWrapper,
    });

    await waitForElementToBeRemoved(() => screen.getByText("Loading..."));
    expect(component).toMatchSnapshot();
  });

  it("should be disabled if isLoading flag is passed ", async () => {
    render(<BreedPicker isLoading={true} />, { wrapper: TestsWrapper });

    const dropdown = screen.getByTestId("dropdown");
    const user = userEvent.setup();
    await user.click(dropdown);

    const menu = screen.queryByTestId("dropdown-options");
    expect(menu).toBeNull();
  });

  it("should preselect 'all'", async () => {
    render(<BreedPicker isLoading={false} />, { wrapper: TestsWrapper });

    await waitForElementToBeRemoved(() => screen.getByText("Loading..."));

    const breed = screen.getByTestId("dropdown-selected");
    expect(breed.textContent).toBe("All");
  });

  it("should show all possible values when expanded", async () => {
    render(<BreedPicker isLoading={false} />, {
      wrapper: TestsWrapper,
    });

    const dropwdown = screen.getByTestId("dropdown");
    const user = userEvent.setup();
    await user.click(dropwdown);

    expect(screen.getByTestId("dropdown-options")).toBeTruthy();

    const options = screen.getAllByTestId("dropdown-option");

    expect(options.length).toEqual(3);
    expect(options[0].textContent).toEqual("All");
    expect(options[1].textContent).toEqual("Abyssinian");
    expect(options[2].textContent).toEqual("Aegean");
  });

  it("should display correct breed based on query param", async () => {
    vi.doMock("@/common/hooks/useQueryParams", () => {
      return {
        useQueryParams: () => {
          return {
            params: { page: "2", breed_ids: "abys", order: "ASC" },
            setQueryParams: () => null,
          };
        },
      };
    });

    const { BreedPicker } = await import("./BreedPicker");
    render(<BreedPicker isLoading={false} />, { wrapper: TestsWrapper });
    await waitForElementToBeRemoved(() => screen.getByText("Loading..."));

    expect(screen.getByTestId("dropdown-selected").textContent).toEqual(
      "Abyssinian"
    );
  });

  it("should set query param (breed_ids and page = 1) when the breed is selected", async () => {
    const setQueryParamsMock = vi.fn(() => null);
    vi.doMock("@/common/hooks/useQueryParams", () => {
      return {
        useQueryParams: () => {
          return {
            params: { page: null, breed_ids: "aege", order: "ASC" },
            setQueryParams: setQueryParamsMock,
          };
        },
      };
    });

    const { BreedPicker } = await import("./BreedPicker");

    render(<BreedPicker isLoading={false} />, { wrapper: TestsWrapper });

    const dropdown = screen.getByTestId("dropdown");
    const user = userEvent.setup();
    await user.click(dropdown);

    const options = screen.getAllByTestId("dropdown-option");
    await user.click(options[2]);

    expect(setQueryParamsMock).toHaveBeenCalledTimes(3);
    expect(setQueryParamsMock).toHaveBeenCalledWith({
      breed_ids: "aege",
      order: "ASC",
      page: null,
    });
  });
});
