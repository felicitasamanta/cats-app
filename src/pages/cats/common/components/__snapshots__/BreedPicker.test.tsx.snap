// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BreedPicker > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="_container_7c5b13"
        >
          <div
            class="_input_7c5b13"
            data-testid="dropdown"
          >
            <div
              class="_selected_7c5b13"
              data-testid="dropdown-selected"
            >
              All
            </div>
            <div
              class="_tools_7c5b13"
            >
              <div
                class="_tool_7c5b13"
              >
                <svg
                  height="20"
                  viewBox="0 0 30 30"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M 2.65625 6.25 L 1.34375 7.75 L 11.34375 16.75 L 12 17.34375 L 12.65625 16.75 L 22.65625 7.75 L 21.34375 6.25 L 12 14.65625 L 2.65625 6.25 z"
                    fill="#000000"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="_container_7c5b13"
      >
        <div
          class="_input_7c5b13"
          data-testid="dropdown"
        >
          <div
            class="_selected_7c5b13"
            data-testid="dropdown-selected"
          >
            All
          </div>
          <div
            class="_tools_7c5b13"
          >
            <div
              class="_tool_7c5b13"
            >
              <svg
                height="20"
                viewBox="0 0 30 30"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M 2.65625 6.25 L 1.34375 7.75 L 11.34375 16.75 L 12 17.34375 L 12.65625 16.75 L 22.65625 7.75 L 21.34375 6.25 L 12 14.65625 L 2.65625 6.25 z"
                  fill="#000000"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
